spec/formatting.md

```diff
@@ -246,65 +247,81 @@ the following steps are taken:
-  the _function_ starting sigil followed by its _identifier_
+  - Otherwise: the U+FFFD REPLACEMENT CHARACTER `�`
 
-  > Examples: `:platform`, `+tag`, `-tag`
+    > Examples:
+    > In any context, `{@reserved\\}` and `{? reserved ?}` both resolve to the _fallback value_ `�`.
 
-- Otherwise: The U+FFFD REPLACEMENT CHARACTER `�` character
+- _expression_ with _literal_ _operand_:
+  U+007C VERTICAL LINE `|`
+  followed by the value of the _literal_
+  with escaping applied to U+005C REVERSE SOLIDUS `\` and U+007C VERTICAL LINE `|`,
+  and then by U+007C VERTICAL LINE `|`.
+  The same representation is used for both _quoted_ and _unquoted_ values.
 
-_Option_ identifiers and values are not included in the _fallback value_.
+  > Examples:
+  > In a context where `:func` fails to resolve,
+  > `{42 :func}` resolves to the _fallback value_ `|42|` and
+  > `{|C:\\| :func}` resolves to the _fallback value_ `|C:\\|`.
+  > In any context, `{|| @reserved}` resolves to the _fallback value_ `||`.
 
-When an error occurs in an _expression_ with a _variable_ _operand_
-and the _variable_ refers to a local _declaration_,
-the _fallback value_ is formatted based on the _expression_
-on the right-hand side of the _declaration_,
-rather than the _expression_ in the _selector_ or _pattern_.
+- _expression_ with _variable_ _operand_ (with or without an _annotation_):
+  - If the _variable_ refers to a local _declaration_:
+    the _value_ to which it resolves (which may already be a _fallback value_)
 
-> For example,
-> in a context in which the function `:func` fails to resolve,
-> attempting to format either of the following messages:
->
-> ```
-> {{
-> local $var = {|horse| :func}
-> {{The value is {$var}.}}
-> }}
-> ```
->
-> ```
-> {{
-> local $var = {|horse|}
-> {{The value is {$var :func}.}}
-> }}
-> ```
->
-> would in both cases result in the _pattern_ _expression_
-> resolving to a _fallback value_ of `|horse|`.
+    > Examples:
+    > In a context where `:func` fails to resolve,
+    > the _pattern_'s _expression_ in `.local $var={|val|} {{{$val :func}}}`
+    > resolves to the _fallback value_ `|val|` and the message formats to `{|val|}`.
+    > In a context where `:now` fails to resolve but `:datetime` does not,
+    > the _pattern_'s _expression_ in
+    > ```
+    > .local $t = {:now format=iso8601}
+    > .local $pretty_t = {$t :datetime}
+    > {{{$pretty_t}}}
+    > ```
+    > (transitively) resolves to the _fallback value_ `:now` and
+    > the message formats to `{:now}`.
+
+  - Otherwise: U+0024 DOLLAR SIGN `$` followed by the _name_ of the _variable_
+
+    > Examples:
+    > In a context where `$var` fails to resolve, `{$var}` and `{$var :number}` and `{$var @reserved}`
+    > all resolve to the _fallback value_ `$var`.
+    > In a context where `:func` fails to resolve,
+    > the _pattern_'s _expression_ in `.input $arg {{{$arg :func}}}`
+    > resolves to the _fallback value_ `$arg` and
+    > the message formats to `{$arg}`.
```

Comment:
I think this whole block should be moved to about line 262 so that the block is effectively structured:

> The fallback value depends on the contents of the expression:
> - expression with literal operand
>   - emit the literal
> - expression with variable operand
>   - emit the variable
> - expression with no operand but with a function
>   - emit the function name
> - expression with no operand but with reserved/private
>   - if sigil+identifier, emit that, else fall through to:
> - otherwise emit the logo

Reply:
That also would be a behavior change, because not every variable operand is represented as a variable (specifically, a variable referencing a `.local` declaration is transparently treated as its resolved value).

Reply:
Fair enough, but the above is a short representation of the whole. Do you disagree that reordering the text would make it clearer?

> The fallback value depends on the contents of the expression:
> - expression with literal operand
>   - emit the literal
> - expression with variable operand
>   - if the variable references a `.local` emit the resolved value
>   - else emit the variable
> - expression with no operand but with a function
>   - emit the function name
> - expression with no operand but with reserved/private
>   - if sigil+identifier, emit that, else fall through to:
> - otherwise emit the logo

Reply:
I think a single-level list of cases with no nested branches _is_ clearer, so I've updated accordingly (basically just promoting "variable referencing `.local`" vs. "variable not referencing `.local`" into distinct cases).